### PR TITLE
Update SAR check for SA Token

### PR DESCRIPTION
**Pull request by jmontleon**
We're looking to stop using the mig serviceAccount which is over privileged. This isn't a definitive check of every permission required, but should be a good start to determining if a SA is properly privileged to be used for a migCluster. We may have to / want to find a sane way to check for everything down the road, although if I'm following how this is done it would be a huge number of checks .

### CODE REVIEW

**Comment by jmontleon** [comment]
Looks like the Linux failure is happening on other commits and not specific to my PR? Any insight?

**Comment by eriknelson** [comment]
@Danil-Grigorev I'm testing that right now, I'd like to get this one merged asap if it looks good to controller folks. I will report back with my testing.

**Comment by jmontleon** [comment]
@eriknelson not sure why you're having trouble:

I added some logging:
```
  log.Info("i##########################################################")
  log.Info(fmt.Sprintf("migrationSar: %t", migrationSar.Status.Allowed))
  log.Info(fmt.Sprintf("veleroSar: %t", veleroSar.Status.Allowed))
  log.Info("i##########################################################")
  if !migrationSar.Status.Allowed || !veleroSar.Status.Allowed {
    cluster.Status.SetCondition(migapi.Condition{
      Type:     SaTokenNotPrivileged,
      Status:   True,
      Reason:   Unauthorized,
      Category: Critical,
      Message:  SaTokenNotPrivilegedMessage,
    })
  } 
```

```
# oc get po migration-controller-5668cf7dc9-kstlz -o yaml| grep image:
    image: quay.io/jmontleon/mig-controller:gvr
    image: quay.io/jmontleon/mig-controller:gvr
    image: quay.io/jmontleon/mig-controller:gvr
    image: quay.io/jmontleon/mig-controller:gvr

```

```
{"level":"info","ts":1584657657.4911134,"logger":"cluster|54dlt","msg":"i##########################################################"}
{"level":"info","ts":1584657657.4912024,"logger":"cluster|54dlt","msg":"migrationSar: true"}
{"level":"info","ts":1584657657.491217,"logger":"cluster|54dlt","msg":"veleroSar: true"}
{"level":"info","ts":1584657657.4912236,"logger":"cluster|54dlt","msg":"i##########################################################"}

```
$ oc sa get-token migration-controller
eyJhbGciOiJS....CcMFTwccHEt61SPlu8Ag
```
```
# oc get secret -o yaml ocp4
apiVersion: v1
data:
  saToken: ZXlK...OEFn
kind: Secret
metadata:
  creationTimestamp: "2020-03-18T21:32:17Z"
  name: ocp4
  namespace: openshift-migration
  resourceVersion: "14260358"
  selfLink: /api/v1/namespaces/openshift-migration/secrets/ocp4
  uid: f0f47f11-695f-11ea-9965-5254009f7cb2
type: Opaque

[root@192 ~]# echo ZXlK...OEFn | base64 -d
eyJhbGciOiJS...CcMFTwccHEt61SPlu8Ag


image

**Comment by eriknelson** [comment]
Could be my old clusters. I'll retry and update. 

⁣Get BlueMail for Android ​

**Comment by eriknelson** [comment]
This needs to be merged simultaneously with https://github.com/konveyor/mig-operator/pull/259

File: pkg/controller/migcluster/validation.go [hunk 1]
```diff
@@ -226,11 +226,23 @@ func (r *ReconcileMigCluster) validateSaTokenPrivileges(cluster *migapi.MigClust
...
 	// check for access to all verbs on all resources in all namespaces
-	// in order to determine if the service account has cluster-admin
-	sar := auth.SelfSubjectAccessReview{
+	// in the migration.openshift.io and velero.io groups in order to
+	// determine if the service account has sufficient permissions
+	migrationSar := auth.SelfSubjectAccessReview{
 		Spec: auth.SelfSubjectAccessReviewSpec{
 			ResourceAttributes: &auth.ResourceAttributes{
-				Group:    "*",
+				Group:    "migration.openshift.io",
```
Comment (jortel) [inline pkg/controller/migcluster/validation.go]: By not including other groups such as:
- "" core (pods etc)
- "apps"

Won't this defeat the purpose of the validation .. To ensure the token is a cluster admin?

Comment (djwhatle) [reply pkg/controller/migcluster/validation.go]: @jortel we are moving away from requiring cluster-admin and moving towards requiring the exact permissions needed by controller. This is a step towards checking the permissions on a more granular level. 

I agree with you that we need more permissions than are covered by these SAR checks.

Comment (jortel) [reply pkg/controller/migcluster/validation.go]: Understood.  Meant cluster-admin figuratively.

**Pull request merged**